src/tasks/base.py

```diff
@@ -0,0 +1,163 @@
+from typing import Tuple, List, Optional
+from abc import ABC, abstractmethod
+
+import torch
+from pytorch_lightning import LightningModule
+from omegaconf import DictConfig
+
+from src.constructor.constuctor import Constructor
+
+
+class BaseTask(LightningModule, ABC):
+    """An abstract class that represent main methods of tasks."""
+
+    def __init__(self, hparams: DictConfig):
+        """Init BaseTask.
+
+        Args:
+            hparams: Hyperparameters that set in yaml file.
+        """
+        super().__init__()
+        self.__constructor = Constructor(hparams)
+        self._metric_manager = self.__constructor.configure_metrics_manager()
+        self._criterion = self.__constructor.configure_losses()
+        self._hparams = self.__constructor.hparams
+        self.__input_shapes = self._hparams.input_shapes
+        self.example_input_array = [torch.rand(1, *shape) for _, shape in self.__input_shapes.items()]
+
+    @abstractmethod
+    def forward(self, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def forward_with_gt(batch: dict) -> dict:
+        pass
+
+    @abstractmethod
+    def training_step(self, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def validation_step(self, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def test_step(self, *args, **kwargs):
+        pass
+
+    def on_train_start(self) -> None:
+        # TODO check and load checkpoint
+        pass
+
+    def on_test_start(self) -> None:
+        # TODO check and load checkpoint
+        pass
+
+    def training_epoch_end(self, outputs: torch.tensor) -> None:
+        """Is calling at the end of the training epoch with the outputs of all training steps."""
+        avg_loss = torch.stack([x['loss'] for x in outputs]).mean()
+        self.log('train/loss', avg_loss, on_step=False, on_epoch=True)
```

Comment:
It must be overwrite, bcause JoitLoss return avg_loss, with tagged_loss_dict (key - name_loss, value - loss_value). It needed because we want log every loss separatly too.